Fix false positive unit tests

arjunpola opened the pull request:
Fixes an issue where the assertion exception is
thrown in child coroutine scope instead of parent
scope when running tests in SearchStoriesUseCaseTest.kt

ataulm commented on core/src/main/java/io/plaidapp/core/designernews/domain/SearchStoriesUseCase.kt:
```diff
@@ -37,29 +38,33 @@ class SearchStoriesUseCase @Inject constructor(
 ) {
     private var parentJob = Job()
     private val scope = CoroutineScope(dispatcherProvider.main + parentJob)
+    private val exceptionHandler: CoroutineExceptionHandler = CoroutineExceptionHandler { _, e ->
```
Is it necessary to include a `CoroutineExceptionHandler` in the scope for the Job to be cancelled when an exception is thrown? If not, I think this can be removed.

arjunpola replied:
No not necessary, added it to prevent the exception from being thrown twice. But, it would consume other exceptions if we do not check for them using the job.

Will remove it.

ataulm commented on core/src/main/java/io/plaidapp/core/designernews/domain/SearchStoriesUseCase.kt:
```diff
@@ -37,29 +38,33 @@ class SearchStoriesUseCase @Inject constructor(
     private val parentJobs = mutableMapOf<String, Job>()
 
-    operator fun invoke(query: String, page: Int, callback: LoadSourceCallback) {
+    operator fun invoke(query: String, page: Int, callback: LoadSourceCallback): Job? {
```
`LoadSourceCallback` is used to abstract the concept of coroutines from the use case. Adding `Job?` as a return type exposes this concept to clients of this class.

I'm not sure adding this just for the test is worth it.

arjunpola replied:
Sure will remove this too.

ataulm commented on core/src/test/java/io/plaidapp/core/designernews/domain/SearchStoriesUseCaseTest.kt:
```diff
@@ -39,6 +40,7 @@ import java.util.GregorianCalendar
 /**
  * Tests for [SearchStoriesUseCase] mocking the dependencies.
  */
+@InternalCoroutinesApi
```
[This annotation shouldn't be used outside of the `kotlinx.coroutines`](https://github.com/Kotlin/kotlinx.coroutines/blob/master/common/kotlinx-coroutines-core-common/src/Annotations.kt#L30-L32) - we should avoid using `getCancellationException()` since it's not part of the (intended) public (use) API.

arjunpola replied:
Will remove. This was needed for getCancelableException. But, will look for an alternative way of fixing this.

arjunpola commented:
Thank you @ataulm for reviews and feedback :)

I also tried another approach of plumbing an optional coroutine context along with dispatchers. Using the context when provided can make sure that exceptions are propogated to parent scope.

Also, I like the idea of having a fake implementation of LoadSourceCallback, will explore that too 👍

arjunpola commented:
Realized the mistake in this approach. I have another patch with the callback approach.

ataulm commented on core/src/test/java/io/plaidapp/core/designernews/domain/AwesomeTestLoadSourceCallback.kt:
```diff
@@ -0,0 +1,43 @@
+    val error: String? = null
+)
+
+class AwesomeTestLoadSourceCallback : LoadSourceCallback {
```
🙈 My bad on the name.

How about splitting it into two? We never expect both callbacks to be called in the same execution flow.

It can also be moved to the `io.plaidapp.core.data` package (in the test sourceset), to match `LoadSourceCallback` since this isn't specific to `designernews`.

```kotlin
class SourceLoadedCallback : LoadSourceCallback {

    var params: Triple<List<PlaidItem>?, Int, String>? = null

    override fun sourceLoaded(result: List<PlaidItem>?, page: Int, source: String) {
        params = Triple(result, page, source)
    }

    override fun loadFailed(source: String) {
    }
}

class LoadFailedCallback : LoadSourceCallback {

    var source: String? = null

    override fun loadFailed(source: String) {
        this.source = source
    }

    override fun sourceLoaded(result: List<PlaidItem>?, page: Int, source: String) {
    }
}
```

ataulm commented on core/src/test/java/io/plaidapp/core/designernews/domain/LoadStoriesUseCaseTest.kt:
```diff
@@ -82,52 +78,33 @@ class LoadStoriesUseCaseTest {
 
         // The correct callback was called
-        assertTrue(sourceLoadingFailed)
+        assertEquals(SourcesRepository.SOURCE_DESIGNER_NEWS_POPULAR, callback.failedValue())
```
```kotlin
// Then fails to load source
val sourceLoadingFailed = callback.failedValue() != null
assertTrue(sourceLoadingFailed)
```

arjunpola commented:
> I'll leave the rest for one of the project collaborators to review. Thanks again for picking up the issue ☺️

I have addressed all the reviews in this patch. Thank you so much @ataulm for all the feedback 😃

keyboardsurfer commented:
LGTM once @ataulm's remaining comments are addressed.

arjunpola commented:
> Please address & resolve remaining comments by @ataulm. LGTM otherwise.

My bad, did not realize I missed few.
